### DISCUSSION TITLE
Server::readClientRequest not fully reading request

### DISCUSSION THREAD

**Discussion by NickMous** [deployphp/deployer#3983] [Bugs]
### Deployer Version

v8.0.0

### Target OS

Debian 12/MacOS Sequoia

### Which PHP version are you using?

PHP 8.4

### Content of deploy.php or deploy.yaml

```php
<?php
namespace Deployer;

require 'recipe/laravel.php';

// Config

set('repository', 'git@github.com:NickMous/<proj>.git');

add('shared_files', []);
add('shared_dirs', []);
add('writable_dirs', []);

// Hosts
foreach (range(1, 5) as $i) {
    host('<proj>0' . $i)
        ->setHostname('<ip>')
        ->setRemoteUser('deployer')
        ->setPort(<port>)
        ->setDeployPath('/var/www/<proj>/0' . $i);
}

// Hooks

after('deploy:failed', 'deploy:unlock');
```

### Steps to reproduce

Just deploy!

**Comment by NickMous** [comment]
### Discussed in https://github.com/deployphp/deployer/discussions/3981

<div type='discussions-op-text'>

<sup>Originally posted by **NickMous** December 18, 2024</sup>
Hiya there, I'm trying to set up a new project with deployer 8, and I'm getting a JsonException the whole time, without any useful information.
Can anyone help me debug this, if I'm the problem or the script is?

It's a fresh laravel application.

```
[<project-stag>] run cd /var/www/<project>/01 && (ln -nfs --relative releases/6 /var/www/<project>/01/release)
done on <project-stag>
  ⠼
In Server.php line 190:

  [JsonException (3)]
  Control character error, possibly incorrectly encoded


Exception trace:
  at /var/www/html/vendor/deployer/deployer/src/Executor/Server.php:190
 json_decode() at /var/www/html/vendor/deployer/deployer/src/Executor/Server.php:190
 Deployer\Executor\Server->parseRequest() at /var/www/html/vendor/deployer/deployer/src/Executor/Server.php:142
 Deployer\Executor\Server->handleClientRequests() at /var/www/html/vendor/deployer/deployer/src/Executor/Server.php:71
 Deployer\Executor\Server->run() at /var/www/html/vendor/deployer/deployer/src/Executor/Master.php:230
 Deployer\Executor\Master->runTask() at /var/www/html/vendor/deployer/deployer/src/Executor/Master.php:110
 Deployer\Executor\Master->run() at /var/www/html/vendor/deployer/deployer/src/Command/MainCommand.php:135
 Deployer\Command\MainCommand->execute() at /var/www/html/vendor/symfony/console/Command/Command.php:279
 Symfony\Component\Console\Command\Command->run() at /var/www/html/vendor/symfony/console/Application.php:1076
 Symfony\Component\Console\Application->doRunCommand() at /var/www/html/vendor/symfony/console/Application.php:342
 Symfony\Component\Console\Application->doRun() at /var/www/html/vendor/symfony/console/Application.php:193
 Symfony\Component\Console\Application->run() at /var/www/html/vendor/deployer/deployer/src/Deployer.php:289
 Deployer\Deployer::run() at /var/www/html/vendor/deployer/deployer/bin/dep:103
 include() at /var/www/html/vendor/bin/dep:119

deploy [--tag TAG] [--revision REVISION] [--branch BRANCH] [-o|--option OPTION] [-l|--limit LIMIT] [--no-hooks] [--plan] [--start-from START-FROM] [--log LOG] [--profile PROFILE] [--] [<selector>...]

Failed to execute command vendor/bin/dep deploy -vvv: exit status 3
```
</div>

### Additions after some digging

In `src\Executor\Server.php::readClientRequest()` the socket is being read until we have 2 new lines.
Somehow in the socket there are two new lines, but it's not done reading.

I see that in the first chunk that we retrieve we get the headers. Maybe we can get the header `Content-Length` first, and do something with that?

#### Pastebins before I delete the files:
[debug0](https://pastebin.com/4DJSSvQY)
[debug1](https://pastebin.com/UddVG5ik)
[debug2](https://pastebin.com/Ps7Aik4d)

**Reply by antonmedv** [reply]
> Maybe we can get the header Content-Length first, and do something with that? 

This seems reasonable.

**Reply by NickMous** [reply]
@antonmedv I can maybe somewhere put some time into it to try and make it work, because this is the only thing stopping me from trying out version 8 right now.
Let me know if you'd like some effort from me!

**Reply by antonmedv** [reply]
Sure. Go ahead. I will try to debug it after holidays.
